### ISSUE TITLE
Adding support for getIndexedKeys()

### ISSUE THREAD

**Issue by jbmusso** [inolen/titan-node#3]
Hi there,

As far as I understand, titan-node doesn't support (yet) the Titan specific `getIndexedKeys()` method (see [Java source code](https://github.com/thinkaurelius/titan/blob/eb5b7c6d95ee752ba7e9643c1445452ed64d0fff/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/blueprints/TitanBlueprintsTransaction.java#L227)).

I was thinking about diving into titan-node source code, modify https://github.com/inolen/titan-node/blob/master/lib/titan-graph-wrapper.js and then sending a pull request. However, I'm unsure yet how to implement it properly so I felt I should open a discussion first.

Basically, I'm looking for a way to avoid the re-creation of already-existing keys while booting up an application (circumventing [current temporary Titan limitations](https://github.com/thinkaurelius/titan/wiki/Titan-Limitations#key-index-must-be-created-prior-to-key-being-used)). When using @barcrab's grex (https://github.com/entrendipity/grex), I check for the existence of keys before attempting to add any. I'm trying to make my lib support both gremlin-node and grex.

Cheers,

**Comment by jbmusso** [comment]
For some reasons I assumed `getIndexedKeys()` to be Titan specific, which it is not ([See Tinkerpop 2.4.0 Java doc](http://www.tinkerpop.com/docs/javadocs/blueprints/2.4.0/com/tinkerpop/blueprints/util/wrappers/readonly/ReadOnlyKeyIndexableGraph.html#getIndexedKeys%28java.lang.Class%29). Unless there's a specific implementation for this in Titan (which I doubt), we might close this issue and rather open a new one in [gremlin-node](https://github.com/inolen/gremlin-node/issues?state=open).

**Comment by inolen** [comment]
I believe you should be able to use TitanGraphWrapper's getType / makeKey / makeLabel to support this functionality.

**Comment by jbmusso** [comment]
Thanks, I'll have a look into it and let you know.

What I'm trying to do is to batch-create (simultaneously) all missing types only. When using grex over the REST endpoint, I used to do it by fetching indexed keys (one call to `getIndexedKeys()`), looping then `make()`ing missing types in a second function call.

I was thinking that retrieving an array of key names in a single function call would help, using either `getIndexedKeys()` or `getTypes()`, which are currently missing in titan-node.

It might be useful to use Titan's [`getTypes()`](http://thinkaurelius.github.io/titan/javadoc/0.4.1/com/thinkaurelius/titan/core/TitanGraph.html#getTypes(java.lang.Class) which is, as far as I understand Titan's universe, a close cousin to `getIndexedKeys()`. However, node-titan currently only supports `getType()` (singular). Note that `getTypes()` takes a `TitanKey.class` or `TitanLabel.class` (according to [Type Definition Overview](https://github.com/thinkaurelius/titan/wiki/Type-Definition-Overview)), both of which are not currently exposed in titan-node's ClassTypes object.

Speaking of ClassTypes, I'm also experiencing issues using them while working with types. Currently, it seems that `g.ClassTypes.<anything>.class` returns `undefined`.

Cheers,

**Comment by inolen** [comment]
Gotcha.

Where did you get `g.ClassTypes.<type>.class`? That use to be correct, but now everything lays at `g.ClassTypes.<type>`

**Comment by jbmusso** [comment]
I read https://github.com/inolen/titan-node/issues/1 and thought using `.class` was still good to go (ie. mimic Java syntax as close as possible).

On a side note, I'd say allowing `.class` is convenient and allows for a closer 1:1 mapping with the many Java/Groovy examples out there (people may expect this syntax to be valid in a Node.js environment). But hey, it's totally your call (and your lib now :wink:).

**Comment by jbmusso** [comment]
Hi Inolen,

Any update on this by chance?

Thanks, cheers,

**Comment by inolen** [comment]
I'm not sure where exactly this is at.

I wasn't planning on re-introducing `.class`.  If you'd like to open a PR for a getTypes implementation I'd gladly review it.
